### PR TITLE
[MXNet-1343][Fit API]Add CNN integration test for fit() API

### PR DESCRIPTION
**Pull request by abhinavs95**
## Description ##
Add nightly integration tests for fit() API using CNN models.
This PR depends on the parent PR for fit() API #14346
JIRA epic: https://issues.apache.org/jira/projects/MXNET/issues/MXNET-1333

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [MXNet-1343](https://issues.apache.org/jira/browse/MXNET-1343)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- test_estimator_cnn_cpu: test on cpu using synthetic data on alexnet, resnet18_v1, FCN
- test_estimator_cnn_gpu: test on gpu using MNIST dataset on resnet18_v1 and verify training accuracy

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

File: ci/docker/runtime_functions.sh [hunk 1]
```diff
@@ -1296,6 +1296,20 @@ nightly_scala_demo_test_cpu() {
     bash bin/run_im.sh
 }
 
+nightly_estimator_cnn_gpu() {
+    set -ex
+    cd /work/mxnet/tests/nightly/estimator
+    export PYTHONPATH=/work/mxnet/python/
+    python tests/nightly/estimator/test_estimator_cnn_gpu.py
```
Comment (karan6181) [inline ci/docker/runtime_functions.sh]: Is the `test_estimator_cnn_gpu.py` file path correct? The current shell location is already inside `/work/mxnet/tests/nightly/estimator` directory.

File: tests/nightly/JenkinsfileForBinaries [hunk 2]
```diff
@@ -106,6 +106,22 @@ core_logic: {
...
+        ws('workspace/estimator-test-cnn-gpu') {
+          utils.unpack_and_init('gpu', mx_lib)
+          utils.docker_run('ubuntu_nightly_gpu', 'nightly_estimator_test_cnn_gpu', true)
+        }
+      }
+    },
+    'estimator: CNN CPU': {
+      node(NODE_LINUX_CPU) {
+        ws('workspace/estimator-test-cnn-cpu') {
+          utils.unpack_and_init('gpu', mx_lib)
```
Comment (karan6181) [inline tests/nightly/JenkinsfileForBinaries]: nit: 'cpu' ??

File: tests/nightly/estimator/test_estimator_cnn_cpu.py [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+        # Define trainer
+        trainer = gluon.Trainer(net.collect_params(), 'sgd', {'learning_rate': lr})
+        # Define estimator
+        est = estimator.Estimator(net=net,
+                                  loss=loss,
+                                  metrics=acc,
+                                  trainers=trainer,
+                                  context=context)
+        # Call fit() to begin training
+        logging_handler = event_handler.LoggingHandler(est, model_name+'_log', model_name+'_log')
```
Comment (karan6181) [inline tests/nightly/estimator/test_estimator_cnn_cpu.py]: LoggingHandler Class constructor says `def __init__(self, estimator, file_name=None, file_location=None, ):`, is `file_location` name correct?(e.g.: alexnet_log/alexnet_log). I think you have to provide `file_location` as a directory path or you can leave it empty since code takes care of it. Thanks!

File: tests/nightly/estimator/test_estimator_cnn_gpu.py [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+    # Define trainer
+    trainer = gluon.Trainer(net.collect_params(), 'sgd', {'learning_rate': lr})
+    # Define estimator
+    est = estimator.Estimator(net=net,
+                              loss=loss,
+                              metrics=acc,
+                              trainers=trainer,
+                              context=context)
+    # Call fit() to begin training
+    logging_handler = event_handler.LoggingHandler(est, model_name+'_log', model_name+'_log')
```
Comment (karan6181) [inline tests/nightly/estimator/test_estimator_cnn_gpu.py]: same as above comment.

**Comment by karan6181** [comment]
@mxnet-label-bot add [Gluon, pr-work-in-progress]

File: ci/docker/runtime_functions.sh (hunk 1 above)
Comment (abhinavs95) [reply ci/docker/runtime_functions.sh]: fixed!

File: tests/nightly/JenkinsfileForBinaries (hunk 2 above)
Comment (abhinavs95) [reply tests/nightly/JenkinsfileForBinaries]: fixed!

File: tests/nightly/estimator/test_estimator_cnn_cpu.py (hunk 3 above)
Comment (abhinavs95) [reply tests/nightly/estimator/test_estimator_cnn_cpu.py]: Thanks for pointing it out. Removing logging handler from integration tests for now as its undergoing changes. It will be covered in unit tests.

File: tests/nightly/estimator/test_estimator_cnn_gpu.py (hunk 4 above)
Comment (abhinavs95) [reply tests/nightly/estimator/test_estimator_cnn_gpu.py]: same as above

File: tests/nightly/estimator/test_estimator_cnn_gpu.py [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+    trainer = gluon.Trainer(net.collect_params(), 'sgd', {'learning_rate': lr})
+    # Define estimator
+    est = estimator.Estimator(net=net,
+                              loss=loss,
+                              metrics=acc,
+                              trainers=trainer,
+                              context=context)
+    # Call fit() to begin training
+    est.fit(train_data=train_data,
+            # val_data=test_data,
```
Comment (roywei) [inline tests/nightly/estimator/test_estimator_cnn_gpu.py]: add validation data here once #14442 merged

File: tests/nightly/estimator/test_estimator_cnn_cpu.py [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+                                                      mx.nd.zeros(shape=(batch_size, 320, 480)))
+            loss = gluon.loss.SoftmaxCrossEntropyLoss(axis=1)
+            net[-1].initialize(init.Constant(bilinear_kernel(num_classes, num_classes, 64)), ctx=context)
+            net[-2].initialize(init=init.Xavier(), ctx=context)
+        else:
+            net = vision.get_model(model_name, classes=10)
+            dataset = gluon.data.dataset.ArrayDataset(mx.nd.random.uniform(shape=(batch_size, 1, 224, 224)),
+                                                      mx.nd.zeros(batch_size))
+            loss = gluon.loss.SoftmaxCrossEntropyLoss()
+            net.initialize(mx.init.MSRAPrelu(), ctx=context)
```
Comment (roywei) [inline tests/nightly/estimator/test_estimator_cnn_cpu.py]: let's use a more common initializer

File: tests/nightly/estimator/test_estimator_cnn_cpu.py [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+
+import numpy as np
+import mxnet as mx
+from mxnet import gluon, init, nd
+from mxnet.gluon.estimator import estimator, event_handler
+from mxnet.gluon.model_zoo import vision
+
+def bilinear_kernel(in_channels, out_channels, kernel_size):
+    '''
+    Bilinear interpolation using transposed convolution
```
Comment (roywei) [inline tests/nightly/estimator/test_estimator_cnn_cpu.py]: reference for this implementation?

File: tests/nightly/estimator/test_estimator_cnn_gpu.py [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+import os
+import sys
+import mxnet as mx
+from mxnet import gluon
+from mxnet.gluon import data
+from mxnet.gluon.estimator import estimator, event_handler
+from mxnet.gluon.model_zoo import vision
+
+def load_data_mnist(batch_size, resize=None, num_workers=None,
+                    root=os.path.join('~', '.mxnet', 'datasets', 'mnist')):
```
Comment (roywei) [inline tests/nightly/estimator/test_estimator_cnn_gpu.py]: remove `root`, let's not keep the data under `~/.mxnet` in nightly tests.

File: tests/nightly/estimator/test_estimator_cnn_cpu.py (hunk 7 above)
Comment (abhinavs95) [reply tests/nightly/estimator/test_estimator_cnn_cpu.py]: Added link to reference code

File: tests/nightly/estimator/test_estimator_cnn_cpu.py (hunk 6 above)
Comment (abhinavs95) [reply tests/nightly/estimator/test_estimator_cnn_cpu.py]: Changed to Xavier initializer

File: tests/nightly/estimator/test_estimator_cnn_gpu.py (hunk 8 above)
Comment (abhinavs95) [reply tests/nightly/estimator/test_estimator_cnn_gpu.py]: Done!

File: tests/nightly/estimator/test_estimator_cnn_gpu.py (hunk 5 above)
Comment (abhinavs95) [reply tests/nightly/estimator/test_estimator_cnn_gpu.py]: added in 810faef

File: tests/nightly/estimator/test_estimator_cnn_cpu.py [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+        trainer = gluon.Trainer(net.collect_params(), 'sgd', {'learning_rate': lr})
+        # Define estimator
+        est = estimator.Estimator(net=net,
+                                  loss=loss,
+                                  metrics=acc,
+                                  trainers=trainer,
+                                  context=context)
+        # Call fit() to begin training
+        est.fit(train_data=train_data,
+                val_data=train_data,
```
Comment (karan6181) [inline tests/nightly/estimator/test_estimator_cnn_cpu.py]: Can you please create another random validation data and pass it here instead of using same `train_data` for both train and validation. Thanks!

Comment (abhinavs95) [reply tests/nightly/estimator/test_estimator_cnn_cpu.py]: Done!

**Comment by abhinavs95** [comment]
@mxnet-label-bot update [Gluon, Test, pr-awaiting-merge]

File: tests/nightly/estimator/test_estimator_cnn_cpu.py [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+        # Get model and initialize, define loss
+        if model_name is 'FCN':
+            num_classes = 21
+            net = FCN(num_classes=num_classes, ctx=context)
+            train_dataset = gluon.data.dataset.ArrayDataset(mx.nd.random.uniform(shape=(batch_size, 3, 320, 480)),
+                                                      mx.nd.zeros(shape=(batch_size, 320, 480)))
+            val_dataset = gluon.data.dataset.ArrayDataset(mx.nd.random.uniform(shape=(batch_size, 3, 320, 480)),
+                                                      mx.nd.zeros(shape=(batch_size, 320, 480)))
+            loss = gluon.loss.SoftmaxCrossEntropyLoss(axis=1)
+            net[-1].initialize(init.Constant(bilinear_kernel(num_classes, num_classes, 64)), ctx=context)
```
Comment (nswamy) [inline tests/nightly/estimator/test_estimator_cnn_cpu.py]: why not get the net from directly from FCN method?, why split the logic?.

File: tests/nightly/estimator/test_estimator_cnn_gpu.py [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+    '''
+    Test estimator by training resnet18_v1 for 5 epochs on MNIST and verify accuracy
+    '''
+    model_name = 'resnet18_v1'
+    batch_size = 128
+    num_epochs = 5
+    input_size = 224
+    lr = 0.001
+    # Set context
+    if mx.context.num_gpus() > 0:
```
Comment (nswamy) [inline tests/nightly/estimator/test_estimator_cnn_gpu.py]: why not use all GPUs?

File: tests/nightly/estimator/test_estimator_cnn_gpu.py [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+
+import os
+import sys
+import mxnet as mx
+from mxnet import gluon
+from mxnet.gluon import data
+from mxnet.gluon.estimator import estimator, event_handler
+from mxnet.gluon.model_zoo import vision
+
+def load_data_mnist(batch_size, resize=None, num_workers=None):
```
Comment (nswamy) [inline tests/nightly/estimator/test_estimator_cnn_gpu.py]: why do you have we need 2 different files?

File: tests/nightly/estimator/test_estimator_cnn_cpu.py [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+    net.add(gluon.nn.Conv2D(num_classes, kernel_size=1),
+            gluon.nn.Conv2DTranspose(num_classes, kernel_size=64, padding=16, strides=32))
+    return net
+
+def test_estimator():
+    '''
+    Test estimator by doing one pass over each model with synthetic data
+    '''
+    models = ['resnet18_v1',
+              'alexnet',
```
Comment (nswamy) [inline tests/nightly/estimator/test_estimator_cnn_cpu.py]: i don't think we need alexnet here since the fit is not very different than resnet.

File: tests/nightly/estimator/test_estimator_cnn_cpu.py (hunk 10 above)
Comment (abhinavs95) [reply tests/nightly/estimator/test_estimator_cnn_cpu.py]: Updated logic, please have a look again

File: tests/nightly/estimator/test_estimator_cnn_gpu.py (hunk 12 above)
Comment (abhinavs95) [reply tests/nightly/estimator/test_estimator_cnn_gpu.py]: combined tests into 1 file

File: tests/nightly/estimator/test_estimator_cnn_cpu.py (hunk 13 above)
Comment (abhinavs95) [reply tests/nightly/estimator/test_estimator_cnn_cpu.py]: removed alexnet

File: tests/nightly/estimator/test_estimator_cnn_gpu.py (hunk 11 above)
Comment (abhinavs95) [reply tests/nightly/estimator/test_estimator_cnn_gpu.py]: The GPU test just runs 5 epochs on MNIST, I think 1 GPU is enough.

File: ci/docker/runtime_functions.sh [hunk 14]
```diff
@@ -1296,6 +1296,20 @@ nightly_scala_demo_test_cpu() {
     bash bin/run_im.sh
 }
 
+nightly_estimator_cnn_gpu() {
+    set -ex
+    cd /work/mxnet/tests/nightly/estimator
+    export PYTHONPATH=/work/mxnet/python/
+    python test_estimator_cnn.py --type gpu
```
Comment (roywei) [inline ci/docker/runtime_functions.sh]: what's the reason to run a python script instead of using nosetest and assert accuracy at the end?

Comment (abhinavs95) [reply ci/docker/runtime_functions.sh]: Since we are executing a single test here, running the python script suffices. We can shift to nosetests if we add more tests in the future.

File: tests/nightly/JenkinsfileForBinaries [hunk 15]
```diff
@@ -106,6 +106,22 @@ core_logic: {
           utils.docker_run('ubuntu_nightly_gpu', 'nightly_tutorial_test_ubuntu_python3_gpu', true, '1500m')
         }
       }
+    },
```
Comment (piyushghai) [inline tests/nightly/JenkinsfileForBinaries]: IMO, This should be in nightly/Jenkinsfile and not the one for Binaries

File: tests/nightly/estimator/test_estimator_cnn.py [hunk 16]
```diff
@@ -0,0 +1,156 @@
...
+                batch_size=1)
+
+def test_estimator_gpu():
+    '''
+    Test estimator by training resnet18_v1 for 5 epochs on MNIST and verify accuracy
+    '''
+    model_name = 'resnet18_v1'
+    batch_size = 128
+    num_epochs = 5
+    if mx.context.num_gpus() > 0:
```
Comment (piyushghai) [inline tests/nightly/estimator/test_estimator_cnn.py]: You should not fallback to mx.cpu here since you want to test on gpu only in this test.

Comment (abhinavs95) [reply tests/nightly/estimator/test_estimator_cnn.py]: I do not want the test to fail if GPU is not available, it should pass on either context.

File: tests/nightly/JenkinsfileForBinaries (hunk 15 above)
Comment (abhinavs95) [reply tests/nightly/JenkinsfileForBinaries]: I think it should work in either Jenkins file, I placed it in binaries since it contained similar existing nightly tests. I will keep your suggestion in mind and update with a follow up PR in case there are any issues with the current setup.

File: tests/nightly/estimator/test_estimator_cnn.py [hunk 17]
```diff
@@ -0,0 +1,156 @@
...
+                              metrics=acc,
+                              trainers=trainer,
+                              context=context)
+    # Call fit()
+    est.fit(train_data=train_data,
+            val_data=test_data,
+            epochs=num_epochs,
+            batch_size=batch_size)
+
+    assert est.train_stats['train_'+acc.name][num_epochs-1] > 0.75
```
Comment (nswamy) [inline tests/nightly/estimator/test_estimator_cnn.py]: why is this only 75%?

File: tests/nightly/estimator/test_estimator_cnn.py (hunk 16 above)
Comment (nswamy) [reply tests/nightly/estimator/test_estimator_cnn.py]: we want it to fail and know if there is an issue, could you please  change it. Also use all the GPUs, so if we run this same code elsewhere we can make use of it without having to change this code.

Comment (nswamy) [reply tests/nightly/estimator/test_estimator_cnn.py]: deleted "Also use all the GPUs, so if we run this same code elsewhere we can make use of it without having to change this code."

Comment (nswamy) [reply tests/nightly/estimator/test_estimator_cnn.py]: ignore the comment on using all GPUs, we can do it as required.

File: tests/nightly/estimator/test_estimator_cnn.py (hunk 17 above)
Comment (abhinavs95) [reply tests/nightly/estimator/test_estimator_cnn.py]: After 5 epochs the training accuracy is ~85%. I can make it more strict by setting it to 80%. What do you think?

File: tests/nightly/estimator/test_estimator_cnn.py (hunk 16 above)
Comment (abhinavs95) [reply tests/nightly/estimator/test_estimator_cnn.py]: Updated context to use GPU
